lib/api/WP_Auth0_Api_Abstract.php

```diff
@@ -180,47 +182,37 @@ protected function set_path( $path ) {
 	 */
 	protected function set_bearer( $scope ) {
 
-		$this->api_token = wp_cache_get( self::CACHE_KEY, WPA0_CACHE_GROUP );
-		if ( ! $this->api_token ) {
-			$this->api_token = $this->options->get( 'auth0_app_token' ); // REMOVE
-		}
+		$api_token = WP_Auth0_Api_Client_Credentials::get_stored_token();
 
-		if ( $this->api_token ) {
-			try {
-				$this->api_token_decoded = $this->decode_jwt( $this->api_token );
-			} catch ( Exception $e ) {
-				// If we can't decode the token, try a client credentials grant below.
-				$this->api_token = null;
-			}
-		}
-
-		// Could not decode the stored API token or none was found so try to get one via API.
-		if ( ! $this->api_token_decoded && $this->api_client_creds instanceof WP_Auth0_Api_Client_Credentials ) {
-			$this->api_token         = $this->api_client_creds->call();
-			$this->api_token_decoded = $this->api_client_creds->get_token_decoded();
+		// No stored API token so need to get a new one.
+		if ( ! $api_token && $this->api_client_creds instanceof WP_Auth0_Api_Client_Credentials ) {
+			$api_token = $this->api_client_creds->call();
 		}
 
 		// No token to use, error recorded in previous steps.
-		if ( ! $this->api_token_decoded ) {
-			wp_cache_delete( self::CACHE_KEY, WPA0_CACHE_GROUP );
+		if ( ! $api_token ) {
 			return false;
 		}
 
+		if ( WP_Auth0_Api_Client_Credentials::check_stored_scope( $scope ) ) {
+			// Scope exists, add to the header and cache.
+			$this->add_header( 'Authorization', 'Bearer ' . $api_token );
+			return true;
+		}
+
 		// API token is missing the required scope.
-		if ( ! $this->api_token_has_scope( $scope ) ) {
-			WP_Auth0_ErrorManager::insert_auth0_error(
-				__METHOD__,
+		WP_Auth0_ErrorManager::insert_auth0_error(
+			__METHOD__,
+			new WP_Error(
+				'insufficient_scope',
 				// translators: The $scope var here is a machine term and should not be translated.
 				sprintf( __( 'API token does not include the scope %s.', 'wp-auth0' ), $scope )
-			);
-			wp_cache_delete( self::CACHE_KEY, WPA0_CACHE_GROUP );
-			return false;
-		}
+			)
+		);
 
-		// Scope exists, add to the header and cache.
-		$this->add_header( 'Authorization', 'Bearer ' . $this->api_token );
-		wp_cache_set( self::CACHE_KEY, $this->api_token, WPA0_CACHE_GROUP );
-		return true;
+		// Delete the stored token so we can try again.
+		WP_Auth0_Api_Client_Credentials::delete_store();
```

Comment:
why would you need to delete the token? What if the user wants to use the stored token for a different API endpoint call?

Reply:
Then it tries again. We don't want the token stored without required scopes or else we can't try again.